File: SCS003_Missing_Null_Check/src/python/Step2locate.py
#!/usr/bin/env python3
"""
step2_locate.py — XPath Sink Locator
--------------------------------------
SCS003: Missing Null Check (CWE-476)

Takes findings from Step 1 and uses XPath queries on the
srcML XML file to confirm the exact sink location.

Supports multiple Juliet scenarios variants:

binary_if:
    Check 1: Find if_stmt at the use line number
    Check 2: Confirm operator is '&' (bitwise) not '&&'
    Check 3: Confirm '->' dereference exists in same if_stmt

char:
    Check 1: Find expr_stmt containing array index at use line
    Check 2: Confirm pointer variable is accessed via index [0]
    Check 3: Confirm NO if_stmt null guard exists before access    

Input:
    tree      -- lxml parsed srcML XML tree
    finding   -- single finding dict from step1_detect
    scenario  -- variant name e.g. "binary_if" or "char"

Output:
    sink result dict with XPath check results
"""

from utils import NAMESPACES, extract_line

# ═════════════════════════════════════════════
# SCENARIO DISPATCHER
# ═════════════════════════════════════════════
def locate(tree, finding: dict, scenario: str = "binary_if") -> dict:
    """
    Dispatches to the correct XPath locator based on scenario.
 
    Args:
        tree:     lxml etree parsed from srcML XML file
        finding:  single finding dict from step1_detect.detect()
        scenario: Juliet variant name ("binary_if" or "char")
 
    Returns:
        Sink result dict with confirmed flag and XPath check results
    """
    if scenario == "binary_if":
        return locate_binary_if(tree, finding)
    elif scenario == "char":
        return locate_char(tree, finding)
    else:
        print(f"  [ERROR] Unknown scenario: {scenario}")
        return None
# ═════════════════════════════════════════════
# VARIANT 1 — binary_if
# Sink: bitwise & in if condition with -> dereference
# ═════════════════════════════════════════════
def locate_binary_if(tree, finding: dict) -> dict:
    """
    Locates sink for the binary_if variant.
 
    Bad pattern:
        if ((ptr != NULL) & (ptr->intOne == 5))
                         ^
                         bitwise & evaluates both sides
                         causing NPD when ptr is NULL
 
    XPath Checks:
        Check 1: if_stmt exists at use line
        Check 2: operator '&' exists (not '&&')
        Check 3: operator '->' exists (pointer dereference)
    """
     
    variable = finding["variable"]
    function = finding["function"]
    used_at  = finding["used_at"]

    if not used_at:
        return None

    # extract line number from first use location
    # format: "filename.c:line:col" -> "line"
    first_use = used_at[0]
    use_line  = extract_line(first_use)

    print(f"\n  Processing  : {variable} in {function}")
    print(f"  {'-' * 40}")
    print(f"  Use location: {first_use}")
    print(f"  Line number : {use_line}\n")

    sink = {
        "variable"          : variable,
        "function"          : function,
        "scenario"          : "binary_if",
        "cwe"               : "CWE-476",
        "use_line"          : use_line,
        "found_if_stmt"     : False,
        "found_bitwise_and" : False,
        "found_dereference" : False,
        "confirmed"         : False
    }

    # ── CHECK 1: Find if_stmt at the use line ──
    # Uses @pos:start on the if_stmt element itself
    xpath1 = (
        f"//src:if_stmt[@pos:start"
        f"[starts-with(.,'{use_line}:')]]"
    )
    if_stmts = tree.xpath(xpath1, namespaces=NAMESPACES)
    sink["found_if_stmt"] = len(if_stmts) > 0

    if sink["found_if_stmt"]:
        print(f"  [PASS] if_stmt found at line {use_line}")
    else:
        print(f"  [FAIL] No if_stmt found at line {use_line}")
        return sink

    # ── CHECK 2: Confirm operator is '&' not '&&' ──
    # Looks for an operator element with exact text '&'
    xpath2 = (
        f"//src:if_stmt[@pos:start"
        f"[starts-with(.,'{use_line}:')]]"
        f"//src:operator[.='&']"
    )
    bitwise_ops = tree.xpath(xpath2, namespaces=NAMESPACES)
    sink["found_bitwise_and"] = len(bitwise_ops) > 0

    if sink["found_bitwise_and"]:
        print(f"  [PASS] Bitwise '&' operator confirmed (not '&&')")
    else:
        print(f"  [FAIL] No bitwise '&' found — may be safe '&&'")

    # ── CHECK 3: Confirm '->' dereference exists ──
    # Looks for '->' operator inside the same if_stmt
    xpath3 = (
        f"//src:if_stmt[@pos:start"
        f"[starts-with(.,'{use_line}:')]]"
        f"//src:operator[.='->']"
    )
    deref_ops = tree.xpath(xpath3, namespaces=NAMESPACES)
    sink["found_dereference"] = len(deref_ops) > 0

    if sink["found_dereference"]:
        print(f"  [PASS] Pointer dereference '->' confirmed")
    else:
        print(f"  [FAIL] No '->' dereference found")

    # ── ALL 3 CHECKS MUST PASS TO CONFIRM SINK ──
    sink["confirmed"] = (
        sink["found_if_stmt"] and
        sink["found_bitwise_and"] and
        sink["found_dereference"]
    )

    return sink

# ═════════════════════════════════════════════
# VARIANT 2 — char
# Sink: direct array access data[0] on NULL pointer
# with no null guard before it
# ═════════════════════════════════════════════
def locate_char(tree, finding: dict) -> dict:
    """
    Locates sink for the char variant.
 
    Bad pattern:
        char * data = NULL;
        printHexCharLine(data[0]);  <- SINK: no null check before access
 
    Good patterns:
        goodG2B: data = "Good"      <- safe source, not NULL
        goodB2G: if (data != NULL)  <- null guard exists before access
 
    XPath Checks:
        Check 1: expr_stmt with function call exists at use line
        Check 2: array index access data[0] on pointer confirmed
        Check 3: NO null guard (if data != NULL) before the access
    """
    variable = finding["variable"]
    function = finding["function"]
    used_at  = finding["used_at"]
 
    if not used_at:
        return None
 
    first_use = used_at[0]
    use_line  = extract_line(first_use)
 
    print(f"\n  [char] {variable} in {function}")
    print(f"  {'-' * 40}")
    print(f"  Use location : {first_use}")
    print(f"  Line number  : {use_line}\n")
 
    sink = {
        "variable"            : variable,
        "function"            : function,
        "scenario"            : "char",
        "cwe"                 : "CWE-476",
        "use_line"            : use_line,
        "found_expr_stmt"     : False,
        "found_array_access"  : False,
        "found_no_null_guard" : False,
        "confirmed"           : False
    }
 
    # ── CHECK 1: Find expr_stmt at the use line ──
    # The sink is a statement like: printHexCharLine(data[0])
    xpath1 = (
        f"//src:expr_stmt[@pos:start"
        f"[starts-with(.,'{use_line}:')]]"
    )
    expr_stmts = tree.xpath(xpath1, namespaces=NAMESPACES)
    sink["found_expr_stmt"] = len(expr_stmts) > 0
 
    if sink["found_expr_stmt"]:
        print(f"  [PASS] expr_stmt found at line {use_line}")
    else:
        print(f"  [FAIL] No expr_stmt found at line {use_line}")
        return sink
 
    # ── CHECK 2: Confirm array index access on the variable ──
    # Looks for: data[0] — name followed by index element
    xpath2 = (
        f"//src:expr_stmt[@pos:start"
        f"[starts-with(.,'{use_line}:')]]"
        f"//src:name[following-sibling::src:index]"
        f"[.='{variable}']"
    )
    array_access = tree.xpath(xpath2, namespaces=NAMESPACES)
    sink["found_array_access"] = len(array_access) > 0
 
    if sink["found_array_access"]:
        print(f"  [PASS] Array access {variable}[0] confirmed")
    else:
        print(f"  [FAIL] No array access found for {variable}")
 
    # ── CHECK 3: Confirm NO null guard before access ──
    # Safe pattern: if (data != NULL) { data[0] }
    # We look for an if_stmt that checks variable != NULL
    # If none exists -> missing null check -> smell confirmed
    xpath3 = (
        f"//src:if_stmt"
        f"//src:name[.='{variable}']"
        f"[following-sibling::src:operator[.='!=']]"
    )
    null_guards = tree.xpath(xpath3, namespaces=NAMESPACES)
    sink["found_no_null_guard"] = len(null_guards) == 0
 
    if sink["found_no_null_guard"]:
        print(f"  [PASS] No null guard found — missing null check confirmed")
    else:
        print(f"  [FAIL] Null guard exists — pointer is safely checked")
 
    # ── ALL 3 CHECKS MUST PASS ──
    sink["confirmed"] = (
        sink["found_expr_stmt"] and
        sink["found_array_access"] and
        sink["found_no_null_guard"]
    )
 
    return sink

# ═════════════════════════════════════════════
# PRINT REPORT
# ═════════════════════════════════════════════

def print_report(sink_results: list):
    """Prints sink location report to terminal."""
    print("\n" + "=" * 60)
    print("  CWE-476 SINK LOCATION REPORT")
    print("=" * 60)
    confirmed = sum(1 for s in sink_results if s and s["confirmed"])
    print(f"  Sinks confirmed: {confirmed} / {len(sink_results)}")
    print("=" * 60)

    for i, s in enumerate(sink_results, 1):
        if not s:
            continue
        scenario = s.get("scenario", "unknown")
        print(f"\n  [{i}] Variable : {s['variable']}")
        print(f"      Function : {s['function']}")
        print(f"      Scenario : {scenario}")
        print(f"      Use line : {s['use_line']}")
        print(f"      CWE      : {s['cwe']}")
        print(f"\n      XPath Checks:")
        
        if scenario == "binary_if":
            print(f"        {'[PASS]' if s.get('found_if_stmt')     else '[FAIL]'}"
                  f"  if_stmt found at line {s['use_line']}")
            print(f"        {'[PASS]' if s.get('found_bitwise_and') else '[FAIL]'}"
                  f"  operator is '&' (bitwise, not '&&')")
            print(f"        {'[PASS]' if s.get('found_dereference') else '[FAIL]'}"
                  f"  '->' dereference exists")
 
        elif scenario == "char":
            print(f"        {'[PASS]' if s.get('found_expr_stmt')     else '[FAIL]'}"
                  f"  expr_stmt found at line {s['use_line']}")
            print(f"        {'[PASS]' if s.get('found_array_access')  else '[FAIL]'}"
                  f"  array access {s['variable']}[0] confirmed")
            print(f"        {'[PASS]' if s.get('found_no_null_guard') else '[FAIL]'}"
                  f"  no null guard before access")
 
        status = "SINK CONFIRMED ⚠️" if s["confirmed"] else "NOT confirmed"
        print(f"\n      {status}")
 
    print("\n" + "=" * 60 + "\n")


# ═════════════════════════════════════════════
# SAVE REPORT
# ═════════════════════════════════════════════
def save_report(sink_results: list, json_file: str, reports_dir: str):
    """
    Saves sink location report as JSON to reports_dir.

    Args:
        sink_results: list of sink dicts from locate()
        json_file:    original input JSON path (used for filename)
        reports_dir:  output directory path
    """
    import os
    import json

    os.makedirs(reports_dir, exist_ok=True)

    base_name   = os.path.basename(json_file).replace(".json", "_sink_report.json")
    report_file = os.path.join(reports_dir, base_name)

    report = {
        "cwe"         : "CWE-476",
        "smell"       : "null-pointer-dereference",
        "source_file" : json_file,
        "total_sinks" : len(sink_results),
        "confirmed"   : sum(1 for s in sink_results if s and s["confirmed"]),
        "results"     : []
    }

    for s in sink_results:
        if not s:
            continue
        report["results"].append({
            "variable"       : s["variable"],
            "function"       : s["function"],
            "scenario"       : s.get("scenario", "unknown"),
            "cwe"            : s["cwe"],
            "use_line"       : s["use_line"],
            "sink_confirmed" : s["confirmed"],
            "xpath_checks"   : {k: v for k, v in s.items()
                                if k.startswith("found_")}
        })

    with open(report_file, "w", encoding="utf-8") as f:
        json.dump(report, f, indent=4)

    print(f"  Sink report saved to: {report_file}")
    return report_file

File: SCS003_Missing_Null_Check/src/cpp/pipeline/Step2locate.cpp
/*
 * Step2locate.cpp — Implementation of Step 2 XPath sink locator
 * ---------------------------------------------------------------
 * SCS003: Missing Null Check (CWE-476) — binary_if and char scenarios
 *
 * Takes findings from Step 1 and runs XPath queries on the srcML
 * XML file to confirm the exact sink location.
 *
 * binary_if:
 *   Check 1: if_stmt exists at use line
 *   Check 2: operator '&' exists (not '&&')
 *   Check 3: operator '->' exists (pointer dereference)
 *
 * char:
 *   Check 1: expr_stmt exists at use line
 *   Check 2: array access variable[0] confirmed
 *   Check 3: NO null guard (if variable != NULL) before access
 */

#include "../include/Step2locate.hpp"
#include <iostream>
#include <fstream>
#include <sys/stat.h>
#include <cstring>
#include <libgen.h>

static const xmlChar* NS_SRC = BAD_CAST "http://www.srcML.org/srcML/src";
static const xmlChar* NS_POS = BAD_CAST "http://www.srcML.org/srcML/position";
static const xmlChar* NS_CPP = BAD_CAST "http://www.srcML.org/srcML/cpp";


std::string extract_line(const std::string& location) {
    size_t last        = location.rfind(':');
    if (last == std::string::npos) return location;
    size_t second_last = location.rfind(':', last - 1);
    if (second_last == std::string::npos) return location;
    return location.substr(second_last + 1, last - second_last - 1);
}

static xmlXPathContextPtr make_xpath_ctx(xmlDocPtr doc) {
    xmlXPathContextPtr ctx = xmlXPathNewContext(doc);
    xmlXPathRegisterNs(ctx, BAD_CAST "src", NS_SRC);
    xmlXPathRegisterNs(ctx, BAD_CAST "pos", NS_POS);
    xmlXPathRegisterNs(ctx, BAD_CAST "cpp", NS_CPP);
    return ctx;
}

static int xpath_count(xmlDocPtr doc, const std::string& expr) {
    xmlXPathContextPtr ctx    = make_xpath_ctx(doc);
    xmlXPathObjectPtr  result = xmlXPathEvalExpression(BAD_CAST expr.c_str(), ctx);

    int count = 0;
    if (result && result->nodesetval)
        count = result->nodesetval->nodeNr;

    xmlXPathFreeObject(result);
    xmlXPathFreeContext(ctx);
    return count;
}


SinkResult locate(xmlDocPtr doc, const json& finding, const std::string& scenario) {
    if (scenario == "binary_if")  return locate_binary_if(doc, finding);
    else if (scenario == "char")  return locate_char(doc, finding);
    else {
        std::cerr << "  [ERROR] Unknown scenario: " << scenario << "\n";
        SinkResult empty{};
        empty.confirmed = false;
        return empty;
    }
}


SinkResult locate_binary_if(xmlDocPtr doc, const json& finding) {
    std::string variable = finding.value("variable", "");
    std::string function = finding.value("function", "");
    std::vector<std::string> used_at = finding.value("used_at", std::vector<std::string>{});

    SinkResult sink{};
    sink.variable          = variable;
    sink.function          = function;
    sink.scenario          = "binary_if";
    sink.cwe               = "CWE-476";
    sink.found_if_stmt     = false;
    sink.found_bitwise_and = false;
    sink.found_dereference = false;
    sink.confirmed         = false;

    if (used_at.empty()) return sink;

    std::string first_use = used_at[0];
    std::string use_line  = extract_line(first_use);
    sink.use_line = use_line;

    std::cout << "\n  Processing  : " << variable << " in " << function << "\n";
    std::cout << "  " << std::string(40, '-') << "\n";
    std::cout << "  Use location: " << first_use << "\n";
    std::cout << "  Line number : " << use_line  << "\n\n";

    std::string xpath1 = "//src:if_stmt[@pos:start[starts-with(.,'" + use_line + ":')]]";
    sink.found_if_stmt = xpath_count(doc, xpath1) > 0;

    if (sink.found_if_stmt)
        std::cout << "  [PASS] if_stmt found at line " << use_line << "\n";
    else {
        std::cout << "  [FAIL] No if_stmt found at line " << use_line << "\n";
        return sink;
    }

    sink.found_bitwise_and = xpath_count(doc, xpath1 + "//src:operator[.='&']") > 0;
    std::cout << "  " << (sink.found_bitwise_and ? "[PASS]" : "[FAIL]") << " Bitwise '&' operator\n";

    sink.found_dereference = xpath_count(doc, xpath1 + "//src:operator[.='->']") > 0;
    std::cout << "  " << (sink.found_dereference ? "[PASS]" : "[FAIL]") << " Pointer dereference '->'\n";

    sink.confirmed = sink.found_if_stmt && sink.found_bitwise_and && sink.found_dereference;
    return sink;
}


SinkResult locate_char(xmlDocPtr doc, const json& finding) {
    std::string variable = finding.value("variable", "");
    std::string function = finding.value("function", "");
    std::vector<std::string> used_at = finding.value("used_at", std::vector<std::string>{});

    SinkResult sink{};
    sink.variable            = variable;
    sink.function            = function;
    sink.scenario            = "char";
    sink.cwe                 = "CWE-476";
    sink.found_expr_stmt     = false;
    sink.found_array_access  = false;
    sink.found_no_null_guard = false;
    sink.confirmed           = false;

    if (used_at.empty()) return sink;

    std::string first_use = used_at[0];
    std::string use_line  = extract_line(first_use);
    sink.use_line = use_line;

    std::cout << "\n  [char] " << variable << " in " << function << "\n";
    std::cout << "  " << std::string(40, '-') << "\n";
    std::cout << "  Use location : " << first_use << "\n";
    std::cout << "  Line number  : " << use_line  << "\n\n";

    std::string xpath1 = "//src:expr_stmt[@pos:start[starts-with(.,'" + use_line + ":')]]";
    sink.found_expr_stmt = xpath_count(doc, xpath1) > 0;

    if (sink.found_expr_stmt)
        std::cout << "  [PASS] expr_stmt found at line " << use_line << "\n";
    else {
        std::cout << "  [FAIL] No expr_stmt found at line " << use_line << "\n";
        return sink;
    }

    sink.found_array_access = xpath_count(doc,
        xpath1 + "//src:name[following-sibling::src:index][.='" + variable + "']") > 0;
    std::cout << "  " << (sink.found_array_access ? "[PASS]" : "[FAIL]") << " Array access " << variable << "[0]\n";

    sink.found_no_null_guard = xpath_count(doc,
        "//src:if_stmt//src:name[.='" + variable + "'][following-sibling::src:operator[.='!=']]") == 0;
    std::cout << "  " << (sink.found_no_null_guard ? "[PASS]" : "[FAIL]") << " No null guard\n";

    sink.confirmed = sink.found_expr_stmt && sink.found_array_access && sink.found_no_null_guard;
    return sink;
}


void print_report(const std::vector<SinkResult>& sink_results) {
    int confirmed = 0;
    for (const auto& s : sink_results)
        if (s.confirmed) confirmed++;

    std::cout << "\n" << std::string(60, '=') << "\n";
    std::cout << "  CWE-476 SINK LOCATION REPORT\n";
    std::cout << std::string(60, '=') << "\n";
    std::cout << "  Sinks confirmed: " << confirmed << " / " << sink_results.size() << "\n";
    std::cout << std::string(60, '=') << "\n";

    int i = 1;
    for (const auto& s : sink_results) {
        std::cout << "\n  [" << i++ << "] Variable : " << s.variable << "\n";
        std::cout << "      Function : " << s.function << "\n";
        std::cout << "      Scenario : " << s.scenario << "\n";
        std::cout << "      Use line : " << s.use_line << "\n";
        std::cout << "      CWE      : " << s.cwe      << "\n";
        std::cout << "\n      XPath Checks:\n";

        if (s.scenario == "binary_if") {
            std::cout << "        " << (s.found_if_stmt     ? "[PASS]" : "[FAIL]") << "  if_stmt at line " << s.use_line << "\n";
            std::cout << "        " << (s.found_bitwise_and ? "[PASS]" : "[FAIL]") << "  operator '&'\n";
            std::cout << "        " << (s.found_dereference ? "[PASS]" : "[FAIL]") << "  '->' dereference\n";
        } else if (s.scenario == "char") {
            std::cout << "        " << (s.found_expr_stmt     ? "[PASS]" : "[FAIL]") << "  expr_stmt at line " << s.use_line << "\n";
            std::cout << "        " << (s.found_array_access  ? "[PASS]" : "[FAIL]") << "  array access " << s.variable << "[0]\n";
            std::cout << "        " << (s.found_no_null_guard ? "[PASS]" : "[FAIL]") << "  no null guard\n";
        }

        std::cout << "\n      " << (s.confirmed ? "SINK CONFIRMED" : "NOT confirmed") << "\n";
    }

    std::cout << "\n" << std::string(60, '=') << "\n\n";
}


void save_report(const std::vector<SinkResult>& sink_results,
                 const std::string& json_file,
                 const std::string& reports_dir) {
    mkdir(reports_dir.c_str(), 0755);

    char* tmp = strdup(json_file.c_str());
    std::string base_name = basename(tmp);
    free(tmp);
    size_t pos = base_name.rfind(".json");
    if (pos != std::string::npos)
        base_name.replace(pos, 5, "_sink_report.json");
    std::string report_file = reports_dir + "/" + base_name;

    int confirmed = 0;
    for (const auto& s : sink_results)
        if (s.confirmed) confirmed++;

    json report;
    report["cwe"]         = "CWE-476";
    report["smell"]       = "null-pointer-dereference";
    report["source_file"] = json_file;
    report["total_sinks"] = sink_results.size();
    report["confirmed"]   = confirmed;
    report["results"]     = json::array();

    for (const auto& s : sink_results) {
        json entry;
        entry["variable"]       = s.variable;
        entry["function"]       = s.function;
        entry["scenario"]       = s.scenario;
        entry["cwe"]            = s.cwe;
        entry["use_line"]       = s.use_line;
        entry["sink_confirmed"] = s.confirmed;

        json checks;
        if (s.scenario == "binary_if") {
            checks["found_if_stmt"]     = s.found_if_stmt;
            checks["found_bitwise_and"] = s.found_bitwise_and;
            checks["found_dereference"] = s.found_dereference;
        } else if (s.scenario == "char") {
            checks["found_expr_stmt"]     = s.found_expr_stmt;
            checks["found_array_access"]  = s.found_array_access;
            checks["found_no_null_guard"] = s.found_no_null_guard;
        }
        entry["xpath_checks"] = checks;
        report["results"].push_back(entry);
    }

    std::ofstream out(report_file);
    out << report.dump(4);
    out.close();

    std::cout << "  Sink report saved to: " << report_file << "\n";
}
